api/src/shop/stripe_event.py

```diff
@@ -92,9 +98,133 @@ def stripe_payment_intent_event(subtype, event):
+
+            end_ts = int(line["period"]["end"])
+            start_ts =int(line["period"]["start"])
+
+            # Divide the timespan down to days
+            days = (end_ts-start_ts)/86400
+            try:
+                new_memberships = add_membership_days(member_id=member_id, span_type=span_type, days=days,creation_reason=f"Subscription invoice {invoice['id']} paid")
+                print(f"New memberships for member {member.member_number}\n{new_memberships}")
+            except Exception as e:
+                print(f"ERROR: {e}")
+        
+        db_session.flush()
+        return
+
+def stripe_customer_event(event_subtype:str, event:any):
+
+    try:
+        subscription = event["data"]["object"]
+        meta = subscription["metadata"]
+        
+        if ("makerspace_user_id" not in meta):
+            print(f"Ignoring customer event {event['id']} without correct metadata")
+            return 
+
+        member_id = int(meta["makerspace_user_id"])
+        member:Member = db_session.query(Member).get(member_id)
+        if (event_subtype == Subtype.SUBSCRIPTION_CREATED):
+            # We end up here if a schedule starts the subscription
+            if (meta['subscription_type'] == SubscriptionTypes.MEMBERSHIP):
+                if (member.stripe_membership_subscription_id != None):
+                    print(
+                        f"WARNING! New subscripton {subscription['id']} will overwrite {member.stripe_membership_subscription_id}")
+
+                member.stripe_membership_subscription_id = subscription['id']
+            elif (meta['subscription_type'] == SubscriptionTypes.LAB):
+                if (member.stripe_labaccess_subscription_id != None):
+                    print(
+                        f"WARNING! New subscripton {subscription['id']} will overwrite {member.stripe_labaccess_subscription_id}")
+                member.stripe_labaccess_subscription_id = subscription['id']
+            pass
+
+        elif (event_subtype == Subtype.SUBSCRIPTION_DELETED):
+            if (meta["subscription_type"] == SubscriptionTypes.LAB):
+                if (member.stripe_labaccess_subscription_id == subscription['id']):
+                    member.stripe_labaccess_subscription_id = None
+                else:
+                    print(f"Ignoring delete notification for subscription {subscription['id']} on {member} since it isn't current. (Current is {member.stripe_labaccess_subscription_id})")
+            elif(meta["subscription_type"] == SubscriptionTypes.MEMBERSHIP):
+                if (member.stripe_membership_subscription_id == subscription['id']):
+                    member.stripe_membership_subscription_id = None
+                else:
+                    print(f"Ignoring delete notification for subscription {subscription['id']} on {member} since it isn't current. (Current is {member.stripe_membership_subscription_id})")
+            db_session.flush()
```

Comment:
Why do we only flush for this branch?

Reply:
I blame this on Python's stupid closures being indent based. :) Flushing in all branches makes more sense.